x/oracle/abci.go

```diff
@@ -17,77 +18,131 @@ func EndBlocker(ctx sdk.Context, k Keeper) {
+	abstainOperatorMap := make(map[string]bool)
+	abstainBallot := voteMap[core.MicroLunaDenom]
+	for _, vote := range abstainBallot {
+		abstainOperatorMap[vote.Voter.String()] = true
+	}
+
+	// Remove abstain key
+	delete(voteMap, core.MicroLunaDenom)
+
+	// Iterate through ballots and update exchange rates; drop if not enough votes have been achieved.
+	for denom, ballot := range voteMap {
+
+		// If denom is not in the whitelist, or the ballot for it has failed, then skip
+		if _, exists := whitelist[denom]; !exists || !ballotIsPassing(ctx, ballot, k, powerMap) {
 			continue
 		}
 
-		// Get weighted median prices, and faithful respondants
-		mod, ballotWinners := tally(ctx, ballot, k)
+		// Get weighted median exchange rates, and faithful respondants
+		ballotMedian, ballotWinningClaims := tally(ctx, ballot, powerMap, params.RewardBand)
+
+		// Set the exchange rate
+		k.SetLunaExchangeRate(ctx, denom, ballotMedian)
 
 		// Collect claims of ballot winners
-		for _, winner := range ballotWinners {
-			key := winner.Recipient.String()
-			claim, exists := claimMap[key]
-			if exists {
-				claim.Weight += winner.Weight
-				claimMap[key] = claim
-			} else {
-				claimMap[key] = winner
+		for _, ballotWinningClaim := range ballotWinningClaims {
+			key := ballotWinningClaim.Recipient.String()
+			prevClaim, exists := winnerMap[key]
+
+			// If the validator is not exist in active set, the one will be excluded from the rewardee list
+			if !exists {
+				continue
 			}
+
+			// Update claim
+			prevClaim.Weight += ballotWinningClaim.Weight
+			winnerMap[key] = prevClaim
+
+			// Increase valid votes counter
+			validVotesCounterMap[key]++
 		}
 
-		// Set price to the store
-		k.SetLunaPrice(ctx, denom, mod)
+		// Emit abci events
 		ctx.EventManager().EmitEvent(
-			sdk.NewEvent(types.EventTypePriceUpdate,
+			sdk.NewEvent(types.EventTypeExchangeRateUpdate,
 				sdk.NewAttribute(types.AttributeKeyDenom, denom),
-				sdk.NewAttribute(types.AttributeKeyPrice, mod.String()),
+				sdk.NewAttribute(types.AttributeKeyExchangeRate, ballotMedian.String()),
 			),
 		)
 	}
 
-	// Convert map to array
-	var claimPool types.ClaimPool
-	for _, claim := range claimMap {
-		claimPool = append(claimPool, claim)
+	//---------------------------
+	// Do miss counting & slashing
+
+	whitelistLen := int64(len(whitelist))
+	for operatorBechAddr, count := range validVotesCounterMap {
+		// Skip abstain operators & valid voters
+		if (count == whitelistLen) || (count == 0 && abstainOperatorMap[operatorBechAddr]) {
+			continue
+		}
+
+		// Increase miss counter
+		operator, _ := sdk.ValAddressFromBech32(operatorBechAddr) // error never occur
+		k.SetMissCounter(ctx, operator, k.GetMissCounter(ctx, operator)+1)
+	}
+
+	// Do slash who did miss voting over threshold and
+	// reset miss counters of all validators at the last block of slash window
+	if core.IsPeriodLastBlock(ctx, params.VotePeriod*params.SlashWindow) {
+		k.SlashAndResetMissCounters(ctx)
```

Comment:
`SlashAndResetMissCounters` logic should sit outside of the keeper; consider decomposing the logic to a separate slashing related file, or to a new function in the handler